File: src/brains/probability_map.clj
(ns brains.probability-map
  (require [clojure.contrib.string :as str]
           [brains.data :as data]))

(defn- inc-map [map path]
  (assoc-in map path (+ 1 (get-in map path 0))))

(defn build [outcome {:keys [fields data outcomes]}]
  (loop [d data p-f {} p-o {} p-of {}]
    (if (empty? d)
      {:count (count data) :p-f p-f :p-o p-o :p-of p-of}
      (let [{label :label values :values} (first d)
            o (outcomes label)]
        (recur (rest d)
               (reduce inc-map p-f values)
               (inc-map p-o [o])
               (reduce inc-map p-of
                       (for [[f v] values]
                         [f v o])))))))

(defn- to-prob [pm p]
  (if p
    (/ p (:count pm))))

(defn feature-values [pm f]
  (keys (f (:p-f pm))))

(defn features [pm]
  (keys (:p-f pm)))

(defn p-o [pm outcome]
  (to-prob pm (get (:p-o pm) outcome)))

(defn p-of [pm outcome [f v]]
  (to-prob pm (get-in (:p-of pm) [f v outcome])))

(defn p-f [pm fv]
  (to-prob pm (get-in (:p-f pm) fv)))

(defn outcomes [pm]
  (keys (:p-o pm)))

(defn most-common-outcome [pm]
  (last (sort-by #(p-o pm %)
                 (outcomes pm))))

File: lib/brains/probability_map.rb
module Brains

# Feature -> class probability mapper.
class ProbabilityMap

  attr_reader :count, :pc, :pf

  def initialize(points = nil)
    @count = 0.0
    @pc = {}
    @pf = {}
    train(points) if points
  end

  def classes
    @pc.keys
  end

  def features
    @pf.keys
  end

  def most_common_class
    @pc.to_a.sort { |a,b| a.last <=> b.last }.last.first
  end

  def getpfc(feature, value, klass)
    @pf.fetch(feature, {}).fetch(value, {})[klass]
  end

  def train(points)
    points.each do |klass,features|
      @count += 1
      @pc[klass] ||= 0
      @pc[klass] += 1
      features.each do |f,v|
        @pf[f] ||= {}
        @pf[f][v] ||= {}
        @pf[f][v][klass] ||= 0
        @pf[f][v][klass] += 1
      end
    end
  end

end

end
